### PR TITLE
Replace deprecated HIP API (SWDEV-318154)

### PR DESCRIPTION
**Pull request by averinevg**
`hipCtxGetCurrent()` and `hipCtxSetCurrent()` were added to better support tensorflow which was using the context API. Since HIP context API were removed from tensorflow codebase (https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/58), these calls can be removed.

Changes:
- `hipCtxGetCurrent()` and `hipCtxSetCurrent()` removed
- `hipHccModuleLaunchKernel()` replaced with `hipExtModuleLaunchKernel()`
- `-Wno-deprecated-declarations` removed
- `MIOPEN_WORKAROUND_ROCM_COMPILER_SUPPORT_ISSUE_30` for https://github.com/RadeonOpenCompute/ROCm-CompilerSupport/issues/30 is not necessary since ROCm 4.3.

### CODE REVIEW

File: src/hip/handlehip.cpp [hunk 1]
```diff
@@ -66,24 +66,20 @@ namespace miopen {
 
 #if MIOPEN_WORKAROUND_ROCM_COMPILER_SUPPORT_ISSUE_30
 namespace {
-void toCallHipInit() __attribute__((constructor(1000)));
-void toCallHipInit() { hipInit(0); }
+static void toCallHipInit() __attribute__((constructor(1000)));
```
Comment (atamazov) [inline src/hip/handlehip.cpp:69]: This is not necessary in unnamed namespace.

File: src/hip/handlehip.cpp [hunk 2]
```diff
@@ -66,24 +66,20 @@ namespace miopen {
...
+static void toCallHipInit() { hipInit(0); }
 } // namespace
 #endif
 
 // Get current context
 // We leak resources for now as there is no hipCtxRetain API
-hipCtx_t get_ctx()
+static hipCtx_t get_ctx()
```
Comment (atamazov) [inline src/hip/handlehip.cpp:76]: Please define local functions in unnamed namespace (instead of static)

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/hip/handlehip.cpp (hunk 2 above)
Comment (atamazov) [reply src/hip/handlehip.cpp:76]: Or keep this as is.

**Review by shurale-nkn — APPROVED** [review]
LGTM

File: cmake/EnableCompilerWarnings.cmake [hunk 3]
```diff
@@ -102,7 +102,7 @@ else()
             endif()
             list(APPEND CMAKE_COMPILER_WARNINGS
                 -Wno-missing-field-initializers
-                -Wno-deprecated-declarations
+                -Wno-global-constructors
```
Comment (atamazov) [inline cmake/EnableCompilerWarnings.cmake:105]: Hmm... Why we need this new suppression?

File: src/hip/handlehip.cpp [hunk 4]
```diff
@@ -58,30 +58,11 @@
 #include <chrono>
 #include <thread>
 
-#define MIOPEN_WORKAROUND_ROCM_COMPILER_SUPPORT_ISSUE_30 (MIOPEN_USE_COMGR && BUILD_SHARED_LIBS)
```
Comment (atamazov) [inline src/hip/handlehip.cpp:61]: Is it so that we no longer need a workaround for https://github.com/RadeonOpenCompute/ROCm-CompilerSupport/issues/30 ?

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/hip/handlehip.cpp (hunk 4 above)
Comment (atamazov) [reply src/hip/handlehip.cpp:61]: [process] @averinevg Let me "resolve" my review comments myself (unless the case is trivial), thanks!

**Review by atamazov — APPROVED** [review]
LGTM!

**Pull request merged**